source/crypto.c

```diff
@@ -41,7 +63,15 @@ PyObject *aws_py_sha256_new(PyObject *self, PyObject *args) {
         return PyErr_AwsLastError();
     }
 
-    return PyCapsule_New(sha256, s_capsule_name_hash, s_hash_destructor);
+    PyObject *capsule = PyCapsule_New(sha256, s_capsule_name_hash, s_hash_destructor);
+
+    /* Why do we return NULL instead of PyErr */
+    if (capsule == NULL) {
+        aws_hash_destroy(sha256);
+        return NULL;
+    }
```

Comment:
Read this: https://docs.python.org/3/extending/extending.html#intermezzo-errors-and-exceptions

then remove comment

the call to `PyErr_AwsLastError()` doesn't actually return an exception. It sets an exception in a global static variable and just returns NULL, which is how python-c-api functions indicate that an error occurred (telling python to look at hte global static variable for details on the error)